File: scripts/test_hnsw.py
#!/usr/bin/env python3
from __future__ import annotations

import os
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, os.fspath(ROOT))

from mori_hnsw import HNSWIndex


def main() -> None:
    index = HNSWIndex.create(dim=4, max_elements=32, ef_search=32)
    index.add(101, [1.0, 0.0, 0.0, 0.0])
    index.add(102, [0.0, 1.0, 0.0, 0.0])
    index.add(103, [0.0, 0.0, 1.0, 0.0])

    hits = index.search([0.99, 0.01, 0.0, 0.0], 2)
    assert hits, "expected search hits"
    assert hits[0].label == 101, f"unexpected best label: {hits[0]}"

    with tempfile.TemporaryDirectory(prefix="mori_hnsw_py_") as tmpdir:
        path = os.path.join(tmpdir, "index.bin")
        index.save(path)
        index.close()

        reloaded = HNSWIndex.load(path, dim=4, ef_search=32)
        hits2 = reloaded.search([0.98, 0.02, 0.0, 0.0], 2)
        assert hits2 and hits2[0].label == 101, f"unexpected reload result: {hits2}"
        print("python-hnsw-ok")

    ip_index = HNSWIndex.create(dim=4, max_elements=16, space="ip", ef_search=16)
    ip_index.add(301, [2.0, 0.0, 0.0, 0.0])
    ip_index.add(302, [1.0, 0.0, 0.0, 0.0])
    ip_hits = ip_index.search([3.0, 0.0, 0.0, 0.0], 2)
    assert ip_hits and ip_hits[0].label == 301, f"unexpected ip result: {ip_hits}"
    assert abs(ip_hits[0].similarity - 6.0) < 1e-5, f"unexpected ip similarity: {ip_hits[0]}"
    with tempfile.TemporaryDirectory(prefix="mori_hnsw_ip_py_") as tmpdir:
        path = os.path.join(tmpdir, "index.bin")
        ip_index.save(path)
        ip_index.close()

        ip_reloaded = HNSWIndex.load(path, dim=4, space="ip", ef_search=16)
        ip_hits2 = ip_reloaded.search([3.0, 0.0, 0.0, 0.0], 2)
        assert ip_hits2 and ip_hits2[0].label == 301, f"unexpected reloaded ip result: {ip_hits2}"
        assert abs(ip_hits2[0].similarity - 6.0) < 1e-5, f"unexpected reloaded ip similarity: {ip_hits2[0]}"
        ip_reloaded.close()


if __name__ == "__main__":
    main()

File: scripts/test_hnsw.lua
#!/usr/bin/env luajit

package.path = package.path .. ";./?.lua;./module/?.lua"

local hnsw = require("module.hnsw")

local function assert_ok(ok, err)
    assert(ok, err)
    return ok
end

local index, err = hnsw.new({
    dim = 4,
    max_elements = 32,
    ef_search = 32,
})
assert(index, err)

assert_ok(index:add(201, {1.0, 0.0, 0.0, 0.0}))
assert_ok(index:add(202, {0.0, 1.0, 0.0, 0.0}))
assert_ok(index:add(203, {0.0, 0.0, 1.0, 0.0}))

local hits = index:search({0.995, 0.005, 0.0, 0.0}, 2)
assert(#hits >= 1, "expected search hits")
assert(hits[1].label == 201, "best hit should be label 201")

local tmp_path = string.format("/tmp/mori_hnsw_lua_%d.bin", os.time())
assert_ok(index:save(tmp_path))
index:close()

local restored, load_err = hnsw.load(tmp_path, {
    dim = 4,
    ef_search = 32,
})
assert(restored, load_err)

local hits2 = restored:search({0.99, 0.01, 0.0, 0.0}, 2)
assert(#hits2 >= 1, "expected reload search hits")
assert(hits2[1].label == 201, "reloaded best hit should be label 201")

local ip_index, ip_err = hnsw.new({
    dim = 4,
    max_elements = 16,
    space = hnsw.SPACE_INNER_PRODUCT,
    ef_search = 16,
})
assert(ip_index, ip_err)

assert_ok(ip_index:add(301, {2.0, 0.0, 0.0, 0.0}))
assert_ok(ip_index:add(302, {1.0, 0.0, 0.0, 0.0}))

local ip_hits = ip_index:search({3.0, 0.0, 0.0, 0.0}, 2)
assert(#ip_hits >= 1, "expected ip search hits")
assert(ip_hits[1].label == 301, "ip best hit should be label 301")
assert(math.abs(ip_hits[1].similarity - 6.0) < 1e-5, "ip similarity should equal dot product")

local ip_tmp_path = string.format("/tmp/mori_hnsw_lua_ip_%d.bin", os.time())
assert_ok(ip_index:save(ip_tmp_path))
ip_index:close()

local ip_restored, ip_load_err = hnsw.load(ip_tmp_path, {
    dim = 4,
    space = hnsw.SPACE_INNER_PRODUCT,
    ef_search = 16,
})
assert(ip_restored, ip_load_err)

local ip_hits2 = ip_restored:search({3.0, 0.0, 0.0, 0.0}, 2)
assert(#ip_hits2 >= 1, "expected reloaded ip search hits")
assert(ip_hits2[1].label == 301, "reloaded ip best hit should be label 301")
assert(math.abs(ip_hits2[1].similarity - 6.0) < 1e-5, "reloaded ip similarity should equal dot product")
ip_restored:close()

print("lua-hnsw-ok")
